File: uid.py
"""Inspired by https://github.com/Mediagone/small-uid"""

import random
import sqlite3
import time
from typing import Optional

MAX_20BIT_INT = 524288


class UID:
    _uid: int

    def __init__(
        self, int: Optional[int] = None, existing: Optional[set["UID"]] = None
    ):
        if existing is None:
            existing = set()
        self._initialize(int)
        while self in existing:
            self._initialize(int)

    def __repr__(self) -> str:
        return str(self._uid)

    def __str__(self) -> str:
        return str(self._uid)

    def __eq__(self, __o: "UID") -> bool:
        return self._uid == __o._uid

    def __hash__(self) -> int:
        return self._uid
    
    def __conform__(self, protocol):
        if protocol is sqlite3.PrepareProtocol:
            return self._uid

    @property
    def int(self):
        return self._uid

    @property
    def hex(self):
        return hex(self._uid)
    
    @classmethod
    def timestamp_ms(cls) -> "int":
        """The current Epoch timestamp, in miliseconds."""
        return int(time.time() * 1000)

    def _initialize(self, i: Optional["int"] = None):
        if i is None:
            time_ms = self.timestamp_ms()
            time_padded = time_ms * 1000000
            rand = random.randrange(MAX_20BIT_INT)
            self._uid = time_padded + rand
        else:
            self._uid = i

File: uid.go
// Mostly a port of https://github.com/Mediagone/small-uid
package uid

import (
	"fmt"
	"math/rand"
	"strconv"
	"time"
)

const max20BitInt = 524288

type UID struct {
	uid int64
}

// Returns a randomly-initialized UID.
func New() *UID {
	return initialize()
}

// Returns a randomly-initialized UID that is guaranteed to not already
// exist in the passed-in map.
func NewWithExisting(existing map[int64]bool) *UID {
	uid := initialize()
	for existing[uid.uid] {
		uid = initialize()
	}
	existing[uid.uid] = true

	return uid
}

func initialize() *UID {
	timeMs := time.Now().UnixMilli()
	rand := rand.Intn(max20BitInt)
	uidStr := fmt.Sprintf("%d%06d", timeMs, rand)

	uid, err := strconv.Atoi(uidStr)
	if err != nil {
		return &UID{uid: -1}
	}

	return &UID{uid: int64(uid)}
}

func (u UID) Int() int64 {
	return u.uid
}

func (u UID) Hex() string {
	return strconv.FormatInt(u.uid, 16)
}
